flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java

```diff
@@ -108,27 +110,30 @@ public class FineGrainedSlotManager implements SlotManager {
     /** ResourceManager's id. */
     @Nullable private ResourceManagerId resourceManagerId;
 
-    /** Executor for future callbacks which have to be "synchronized". */
-    @Nullable private Executor mainThreadExecutor;
+    private final Object lock = new Object();
+    /**
+     * Executor for future callbacks which have to be "synchronized". This field being {@code null}
+     * indicates that the component isn't started.
+     */
+    @GuardedBy("lock")
+    @Nullable
+    private volatile ComponentMainThreadExecutor mainThreadExecutor;
```

Comment:
It's odd that this is both guarded by a lock and volatile.

Reply:
correct, the volatile doesn't make sense here. I have to protected the `mainThreadExecutor` with a lock because of the `runInMainThreadIfStarted` method checks for the main thread executor and executes the Runnable within the lock.